File: apps/ios/Pebbles/Features/Karma/AchievementNotificationService.swift
import Foundation

/// One card of an unlock moment: a badge that just unlocked, with the karma it
/// actually paid (as reported by `check_achievements()`, never the catalog's
/// current value).
struct AchievementMomentCard: Equatable, Sendable, Identifiable {
    /// The badge slug — stable, and the fallback headline if copy is missing.
    let id: String
    let title: String
    let description: String?
    let karmaGranted: Int
}

/// Explicit-fire entry point for the achievement unlock moment (D13).
///
/// Sibling of `KarmaNotificationService`, but a different shape of celebration:
/// where karma flashes a pastille that times out, an unlock opens a chained
/// card per badge that the user taps through. Both are hosted by the same
/// overlay window, which floats above presented sheets — the composer dismisses
/// itself on the very success that fires this, so presenting from the view tree
/// would race that dismissal.
///
/// Only the mutation path celebrates: the screen-open call in `AchievementsView`
/// is the retroactive grant and can return a veteran's whole history at once,
/// so it renders in the grid instead of chaining twenty cards.
@Observable
@MainActor
final class AchievementNotificationService {
    /// The queue being celebrated (empty = nothing showing).
    private(set) var cards: [AchievementMomentCard] = []
    /// Index of the card on screen.
    private(set) var index: Int = 0

    /// Counts feedback fires so tests can assert the guard without CoreHaptics.
    private(set) var hapticTrigger: Int = 0

    private let haptics: HapticsService

    init(haptics: HapticsService = HapticsService()) {
        self.haptics = haptics
    }

    /// The card on screen, or nil when the moment is idle.
    var currentCard: AchievementMomentCard? {
        guard index >= 0, index < cards.count else { return nil }
        return cards[index]
    }

    var isShowingLastCard: Bool {
        index + 1 >= cards.count
    }

    func present(cards: [AchievementMomentCard]) {
        guard !cards.isEmpty else { return }

        // Reuses the karma-earned haptic: same reward register, and a second
        // bespoke waveform for a moment that often fires alongside the karma
        // flash would double-buzz.
        hapticTrigger &+= 1
        haptics.playKarmaEarned()

        self.cards = cards
        self.index = 0
    }

    /// Advances to the next card, ending the moment after the last one.
    func advance() {
        if isShowingLastCard {
            dismiss()
        } else {
            index += 1
        }
    }

    /// Ends the moment immediately — tapping outside or swiping down skips the
    /// rest of the queue. Dismissal is never blocking.
    func dismiss() {
        cards = []
        index = 0
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/karma/AchievementNotificationService.kt
package app.pbbls.android.features.karma

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import app.pbbls.android.services.AchievementRecord
import javax.inject.Inject
import javax.inject.Singleton

/**
 * One card of an unlock moment: a badge that just unlocked, with the karma it
 * actually paid (as reported by `check_achievements()`, never the catalog's
 * current value).
 *
 * [record] carries the catalog row so the card can resolve its localized copy
 * in the composition (`achievementTitle` is `@Composable`); it is null only
 * when a check raced a catalog change, in which case [slug] headlines the card.
 */
data class AchievementMomentCard(
    val slug: String,
    val record: AchievementRecord?,
    val karmaGranted: Int,
)

/**
 * One unlock moment: the card on screen plus where it sits in the queue.
 *
 * The queue is the reason this is not simply a nullable card — a mutation that
 * unlocks three badges shows three cards with "1 of 3" progress, and flattening
 * that to a single card would silently drop the progress UI.
 */
data class AchievementMoment(
    val card: AchievementMomentCard,
    val position: Int,
    val total: Int,
    val isLast: Boolean,
)

/**
 * Explicit-fire entry point for the achievement unlock moment (D13).
 *
 * Sibling of [KarmaNotificationService], but a different shape of celebration:
 * where karma flashes a pastille that times out, an unlock opens a chained card
 * per badge that the user taps through, so there is no auto-dismiss here.
 *
 * Only the mutation path celebrates: the screen-open call in `AchievementsScreen`
 * is the retroactive grant and can return a veteran's whole history at once, so
 * it renders in the grid instead of chaining twenty cards.
 */
@Singleton
class AchievementNotificationService
    @Inject
    constructor() {
        /** The queue being celebrated (empty = nothing showing). */
        var cards: List<AchievementMomentCard> by mutableStateOf(emptyList())
            private set

        /** Index of the card on screen. */
        var index: Int by mutableIntStateOf(0)
            private set

        /** The card on screen, or null when the moment is idle. */
        val currentCard: AchievementMomentCard?
            get() = cards.getOrNull(index)

        val isShowingLastCard: Boolean
            get() = index + 1 >= cards.size

        /**
         * The moment [AchievementMomentOverlay] renders (#852: state, not this
         * service). A getter, not a stored field, so it can never drift from
         * [cards]/[index] — the same reasoning as [currentCard]/[isShowingLastCard].
         */
        val moment: AchievementMoment?
            get() =
                currentCard?.let { card ->
                    AchievementMoment(
                        card = card,
                        position = index + 1,
                        total = cards.size.coerceAtLeast(1),
                        isLast = isShowingLastCard,
                    )
                }

        fun present(cards: List<AchievementMomentCard>) {
            if (cards.isEmpty()) return
            this.cards = cards
            this.index = 0
        }

        /** Advances to the next card, ending the moment after the last one. */
        fun advance() {
            if (isShowingLastCard) dismiss() else index += 1
        }

        /**
         * Ends the moment immediately — tapping the scrim or pressing back skips
         * the rest of the queue. Dismissal is never blocking.
         */
        fun dismiss() {
            cards = emptyList()
            index = 0
        }
    }
